### ISSUE TITLE
SFINAE friendly views

### ISSUE THREAD

**Issue by gnzlbg** [ericniebler/range-v3#432]
Is there an easy way to constrain a function on whether a range composition is a valid expression?

Some views like `view::transform` eagerly trigger an `static_assert` that breaks SFINAE on them. I have to add a layer of indirection to try to SFINAE out before arriving at expression containing views that use `static_assert`.

**Comment by ericniebler** [comment]
It's a trade-off I made to favor good error reporting. And folks do like the error reporting, but I admit it's less than ideal. There isn't a solution in the code as it is. I'm open to suggestions.

**Comment by gnzlbg** [comment]
I think something like this helps:

```c++
template <typename T> struct fail : std::false_type {};

template<typename Rng, CONCEPT_REQUIRES_(Range<Rng>{})>
void foo(Rng&& rng) {
    static_assert(fail<Rng>{}, "message");
}
```

It allows SFINAEing on `foo` as long as the constraints are not met. But if the constraints are met then the `static_assert` triggers.

**Comment by ericniebler** [comment]
I don't think this helps. Take a look at how the view generators are implemented. There are two overloads: one for when the constraints are satisfied and one when they're not. The latter one has the static asserts.

**Comment by badair** [comment]
I took [great pains](https://github.com/badair/callable_traits/blob/master/include/callable_traits/remove_args.hpp) to achieve SFINAE-able errors in my library. The main downside is that your error messages have to be types (and thus underscored and namespaced).

If you can apply the checks early, users will be spared the rest of the template stack in the error message -- although there is some template overhead to make these errors work. [Here](http://melpon.org/wandbox/permlink/xqOGtj4kaG6ljTJI) is an example error message (I probably should've used shorter namespaces for errors). Obviously it's not ideal, but it might be a nice middle ground for users. 

At some point I refactored the error code into it's own little library [here](https://github.com/badair/sfinae_errors), although I believe the CallableTraits version received some unrequited upstream love.

My implementation is SFINAE-able even for template aliases, which (IIRC) required some extra work that wouldn't really be necessary here.
